iam-token-manager/v1.ts

```diff
@@ -0,0 +1,190 @@
+      // 2. request an initial token
+      this.requestToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else if (this.isTokenExpired()) {
+      // 3. refresh a token
+      this.refreshToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(null, this.tokenInfo.access_token);
+    }
+  }
+
+  /**
+   * Request an IAM token using an API key.
+   *
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public requestToken(cb: Function): void {
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: 'Basic Yng6Yng='
+        },
+        form: {
+          grant_type: 'urn:ibm:params:oauth:grant-type:apikey',
+          apikey: this.iamApikey,
+          response_type: 'cloud_iam'
+        }
+      }
+    };
+    sendRequest(parameters, cb);
+  }
+
+  /**
+   * Refresh an IAM token using a refresh token.
+   *
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public refreshToken(cb: Function) {
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: 'Basic Yng6Yng='
+        },
+        form: {
+          grant_type: 'refresh_token',
+          refresh_token: this.tokenInfo.refresh_token
+        }
+      }
+    };
+    sendRequest(parameters, cb);
+  }
+
+  /**
+   * Set a self-managed IAM access token.
+   * The access token should be valid and not yet expired.
+   *
+   * By using this method, you accept responsibility for managing the
+   * access token yourself. You must set a new access token before this
+   * one expires. Failing to do so will result in authentication errors
+   * after this token expires.
+   *
+   * @param {string} iamAccessToken - A valid, non-expired IAM access token
+   * @returns {void}
+   */
+  public setAccessToken(iamAccessToken: string): void {
+    this.userAccessToken = iamAccessToken;
+  }
+
+  /**
+   * Check if currently stored token is expired.
+   * 
+   * Using a buffer to prevent the edge case of the 
+   * token expiring before the request could be made.
+   *
+   * The buffer will be a fraction of the total TTL. Using 80%.
+   *
+   * @private
+   * @returns {boolean}
+   */
+  private isTokenExpired(): boolean {
+    const fractionOfTtl = 0.8;
+    const timeToLive = this.tokenInfo.expires_in;
+    const expireTime = this.tokenInfo.expiration;
+    return expireTime - (timeToLive * (1.0 - fractionOfTtl))
+      < Math.floor(Date.now() / 1000);
```

Comment:
This certainly isn't necessary, but it might help to add a couple of intermediate, explanatory variables here:

```node
const expireTime = this.tokenInfo.expiration;
const refreshTime = expireTime - (timeToLive * (1.0 - fractionOfTtl));
const currentTime = Math.floor(Date.now() / 1000);
return refreshTime < currentTime;
```